src/poetry/repositories/pool.py

```diff
@@ -1,59 +1,79 @@
 from __future__ import annotations
 
+from collections import OrderedDict
+from dataclasses import dataclass
+from enum import IntEnum
+from enum import auto
 from typing import TYPE_CHECKING
 
+from poetry.repositories.abstract_repository import AbstractRepository
+from poetry.repositories.abstract_repository import RepositoryName
 from poetry.repositories.exceptions import PackageNotFound
-from poetry.repositories.repository import Repository
 
 
 if TYPE_CHECKING:
     from poetry.core.constraints.version import Version
     from poetry.core.packages.dependency import Dependency
     from poetry.core.packages.package import Package
 
+    from poetry.repositories.repository import Repository
 
-class Pool(Repository):
+
+class Priority(IntEnum):
+    # The order of the members below dictates the actual priority. The first member has
+    # top priority.
+    DEFAULT = auto()
+    PRIMARY = auto()
+    SECONDARY = auto()
+
+
+@dataclass(frozen=True)
+class PrioritizedRepository:
+    repository: Repository
+    priority: Priority
+
+
+class Pool(AbstractRepository):
     def __init__(
         self,
         repositories: list[Repository] | None = None,
         ignore_repository_names: bool = False,
     ) -> None:
-        super().__init__("poetry-pool")
-
-        if repositories is None:
-            repositories = []
-
-        self._lookup: dict[str, int] = {}
-        self._repositories: list[Repository] = []
-        self._default = False
-        self._has_primary_repositories = False
-        self._secondary_start_idx: int | None = None
+        super().__init__("poetry-repository-pool")
+        self._repositories: OrderedDict[
+            RepositoryName, PrioritizedRepository
+        ] = OrderedDict()
+        self._ignore_repository_names = ignore_repository_names
 
+        repositories = repositories or []
```

Comment:
I'd rather write this as

```python
if repositories is None:
    repositories = []
```

It's more correct as if we use a boolean evaluation, an empty list is false-y. Using the caller's list isn't necessary here, but it doesn't hurt to use the more correct form.